### ISSUE TITLE
Add wait with timeout option on pool

### ISSUE THREAD

**Issue by cocotyty** [garyburd/redigo#316]
when Pool.Wait is true and  network broken , connection will dial timeout or read timeout , pool.active reach MaxActive quickly ,so lots of Get() call will wait at p.cond.Wait(). As time goes by, goroutine leakage will be more and more serious.

Is it possible  Pool can provide wait-timeout future ?

**Comment by cocotyty** [comment]
maybe we can use channel instead of sync.Cond ?

```go
package cond

import "time"

type Cond chan struct{}

func (c Cond) Wait(timeout time.Duration) (success bool) {
	select {
	case _ = <-c:
		return true
	case <-time.After(timeout):
		return false
	}
}

func (c Cond) Signal() {
	select {
	case c <- struct{}{}:
	default:
	}
}

func (c Cond) Broadcast() {
	for {
		select {
		case c <- struct{}{}:
		default:
			return
		}
	}
}

```

**Comment by garyburd** [comment]
Are you saying that many goroutines wait in this scenario, or that there's an actual leak?  A leak is where the routine is blocked forever.

The channel code above is not a drop in replacement for sync.Cond.

I started work replacing sync.Cond with a channel [here](https://github.com/garyburd/redigo/pull/313).  I don't know when I'll have time to complete the work.

**Comment by cocotyty** [comment]
Thank you for your greate project and your  answer.
not actual a leak . when the network is back to normal, goroutine numbers gradually decrease .

the probability of this happening is usually small.

**Comment by cocotyty** [comment]
I want to contribute, but I found it seems almost done :)

**Comment by garyburd** [comment]
I will add a GetContext method.  See the work in progress at https://github.com/garyburd/redigo/blob/867bedcd6726a9863566f6009822cae9899f2842/redis/pool.go#L183-L191.

**Comment by garyburd** [comment]
Fixed by f5d5b3d22ed4d0ea867b75c023f49d80e24a494f.
